plutus-ir/src/Language/PlutusIR/Optimizer/Merge.hs

```diff
@@ -0,0 +1,115 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase       #-}
+-- | Optimization passes for merging independent let-bindings together.
+module Language.PlutusIR.Optimizer.Merge (mergeLets) where
+
+import           Language.PlutusIR
+import qualified Language.PlutusIR.Analysis.Dependencies as Deps
+
+import qualified Language.PlutusCore                     as PLC
+import qualified Language.PlutusCore.Name                as PLC
+
+import           Control.Lens
+import           Control.Monad.Reader
+
+import           Data.List
+import           Data.Maybe
+import qualified Data.Set                                as Set
+
+import qualified Algebra.Graph                           as G
+import qualified Algebra.Graph.ToGraph                   as T
+
+mergeLets
+    :: (PLC.HasUnique (name a) PLC.TermUnique, PLC.HasUnique (tyname a) PLC.TypeUnique)
+    => Term tyname name a
+    -> Term tyname name a
+mergeLets t =
+    let
+        depGraph :: G.Graph Deps.Node
+        depGraph = Deps.runTermDeps t
+    in
+        runReader (mergeTerm t) depGraph
+
+{- Note [Merging lets]
+We can merge independent, non-recursive, let bindings together into a single let binding, which is
+nice since it reduces the nesting of the program.
+
+We can *nearly* do this by just looking at the dependency graph and constructing level sets. However,
+nodes in the same level set may not be mergeable if their let bindings are not adjacent.
+
+So as it stands we need to traverse the term manually, performing local merging.
+-}
+mergeTerm
```

Comment:
This is not shown as a Haddock comment, right? I think, it's useful and small enough to be a Haddock comment.